blib/man1/pcurse.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PCURSE 1"
.TH PCURSE 1 "2011-08-29" "perl v5.12.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
pcurse \- Monitor  Postgres  databases
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 3
\& $ pcurse  [options]   [dbname]   [user]
\& for example:
\& $ pcurse \-h localhost  template1  john  \-\-passwd apple
.Ve
.SS "options"
.IX Subsection "options"
.Vb 5
\&      \-\-passwd      Default is read from ~/.pgpass
\&      \-\-dbname      Default is template1
\&      \-\-host        Default is localhost
\&      \-\-user        Default is getlogin() 
\&      \-\-port        Default is 5432
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fIpcurse\fR\|(1) monitors a Postgres database cluster. 
It shows statistics on tables, indexes, vacuuming, buffers, etc,.
.SH "INVOCATION"
.IX Header "INVOCATION"
.SH "USAGE"
.IX Header "USAGE"
It is too simple:
.PP
1. Select a '\s-1MODE\s0' from the menu.
   To navigate, use arrow keys, the vi keys (j ,k , m ), or \s-1ENTER\s0 .
   \s-1TAB\s0 (or 'l') moves you to the next window.
.PP
2. Most modes requires to select a namespace from the the 'Schemas' listbox.
   Use arrows, vi keys, \s-1ENTER\s0 (or \s-1SPACE\s0, or ' ' ) seals the selection.
   Press 'd' to view more acl permissions on schema
   \s-1TAB\s0 (or 'l') moves you to the button window.
.PP
3. The results are displayed in the bottom window.
   Use arrows, vi keys, \s-1ENTER\s0 (or \s-1SPACE\s0, or ' ' ) seals the selection.
   Pressing 'd' to view more results.
   Ctrl-H displays Help screen
   Ctrl-F appends the contents of the bottom display to /tmp/pcurse.out
   In tables mode Ctrl-L displays the last 20 updates in table
   In tables and database mode Ctrl-A analyses at table or database
   In tables and database mode Ctrl-V analyses at table or database
   Press 'd' to exit from any of the above tables.
   \s-1TBB\s0 (or 'l') takes you back to the starting menu.
.PP
4. Press 'q' to quit the program. Thank you for flying Pcurse!
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.SH "AUTHOR"
.IX Header "AUTHOR"
Ioannis Tambouras, <ioannis@cpan.org>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright (C) 2008 by Ioannis Tambouras
.PP
This library is free software; you can redistribute it and/or modify
it under the same terms of GPLv3


Changes
Revision history for Perl extension Pg::Pcurse.

0.23  Sun Aug 29 01:05:01 2011
       - Regular upkeep: changed installer, changed and added tests

0.22  Wed Jul 23 21:48:23 2008
        - Fixed license entry (to perl)

0.20  Thu Jun 26 02:23:59 2008
        - In vacuum mode, Ctrl-L shows configuration settings
        - In vacuum mode, fixed listing of tables so they are ordered
        - In vacuum mode, Ctrl-T shows table statistics
        - In vacuum mode, changed main display to show vac and analyze
        - Old selction values in listboxes are now sticky ( when same params)

0.19  Sat May 31 02:01:56 2008
	- fsm displays names of relations in other db's by its name
          insted of relfileno
	- In buffers mode, Ctrl-L shows fsm relations
	- In buffers mode, Ctrl-T shows fsm and Checkpoint Stats
	- In buffers mode, 'd' shows fsm pages

0.18  Sun May 25 07:47:58 2008
	- Added index definition
	- In statistics mode, added pg_stats for Ctrl-T and Ctrl-L
	- In tables mode, added pg_statio for Ctrl-T 
	- In rules mode, added pg_rewrite for Ctrl-T 
	- Re-arranged buffers mode. Added filters and more info
          in big display; 'd' shows summary; Ctrl-F shows stats and settings.
	- In settings mode, Ctrl-L views fsm rations

0.17  Thu May 22 23:48:15 2008
        - added format info about pgfouine's log_line_prefix
        - Added the missing Defaults.pm file
        - show num of default buffers TODO
        - Ctrl-F in buffers mode, it displays statistics about checkpoints
        - Fixed indexes mode
        - Added 'dict' mode for text search dictionaries

0.15  Thu May 15 20:37:08 2008
        - minor display fixes
        - Ctrl-H displays Help screen
        - Ctrl-F appends the contents of the bottom display to /tmp/pcurse.out
        - Ctrl-L replaces Shift-d , it displays the last 20 updates in table
        - Ctrl-A replaces Shift-a , it analyses  a table or database
        - Ctrl-V replaces Shift-v , it vacuums   a table or database
        - Ctrl-R replaces Shift-r , it reindexes a table or database
        - vacuum, and auto_vacuum dates are properly displayed
        - In settings mode, added 'changed' that selects only the parameters
          set differently than the default postgresql.conf (mine is from Debian)

0.13  Wed May 14 10:28:45 2008
        - fixed use 5.008 in Makefile.PL
	- Mode databases still works, but under development


0.12  Mon May 12 06:52:36 2008
        - schema table is disable for user mode
        - added extra_desc for settings mode
        - In tables mode, 'D' displays the youngest 20 rows of table
        - In rules mode, beautify rule definition
        - In settings mode, settings are now searchable by context
        - In indexes mode, changed or added more fields
        - In procedures mode, 'D' also displays the src

0.11  Wed May  9 06:48:31 2008
        - Display postmaster start time in some modes
        - Display txid_current in some modes
        - Prevent crashes with restricted databases

0.10  Wed May  8 09:39:46 2008
	- In 'database' mode, 'D' displays count of objects
	- In 'settings' mode, 'D' displays fsm rel and pages
	- less clutter in standard 'database' mode
	- bief summary for buffercache (if available)
	- Schema table is disabled when not needed

0.09  Wed May  7 08:55:42 2008
	- minor code refactoring
	- query refactoring

0.08  Man May  5 06:53:22 2008
	- Attemp to read passwd from ~/.pgpass
	- CLI invocation enhancements

0.07  Fri May  2 06:31:03 2008
	- Renamed mode 'overview' to 'databases'
	- Fixed calculation of row sizes 
	- Added more fields to 'd' tables
	- Display updates after analyze & vacuum 

0.06  Man May  2 05:57:37 2008
	- Lost of micro-features, here are some:
	  'd' on schema display shows permissions
	  'D' on big display shows records
	  'A' attempts to analyze a table or database
	  'V' attempts to vacuum a table or database
	  Bucardo, trigger, and views modes
	  Always display what mode we are int
	  Better vacuum submode
	  Prevent terminal blur on failed startup (invalid db credentials) 

0.05  Man Apr 21 03:52:35 2008
	- changed the blue screen of 'tables' to statsof()
	- Fixed module dependacies

0.04  Sun Apr 20 20:39:33 2008
	- Added 'rules' mode
	- Better lexer for sql stataments
	- added ' ' in keyscan for shcema box
	- added 'h' to un-pop the menu display
	- added acl display for schema box
	- The shcema choice is remembered (when possible) accross 
          round of selections
	- Fixed the default mode back to 'tables'

0.03  Sun Apr 20 04:12:49 2008
	- Added 'views' mode
	- procedure arg type are shown in english names

0.02  Sat Apr 19 19:08:32 2008
	- Corrected module dependencies
	- Fixed usage documentation

0.01  Thu Apr 17 23:06:48 2008
	- original version; created by h2xs 1.23 with options
		-AXn Pg::Pcurse



blib/lib/Pg/Pcurse.pm
# Copyright (C) 2008 Ioannis Tambouras <ioannis@cpan.org>. All rights reserved.
# LICENSE:  GPLv3, eead licensing terms at  http://www.fsf.org .

package Pg::Pcurse;

use v5.8;
use strict;
use warnings;
require Exporter;

use Curses;
use Curses::Widgets;
use Carp::Assert;
use strict;
use Pg::Pcurse::Query0;
use Pg::Pcurse::Query1;
use Pg::Pcurse::Query2;
use Pg::Pcurse::Query3;

our $VERSION = '0.23';

our $opt;

use base 'Exporter';

our @EXPORT = qw( 
	execute_mode       retrieve_context   capital_context
	$opt               retrieve_permit    update_big_display
	analyze            reindex            vacuum   save2file
	stat_of            over3              bufcalc  idx3b
	table2of           table3of           rewrite_of
);

*secondary_listbox = *main::secondary_listbox;
*big_listbox       = *main::big_listbox;
*create_button     = *main::create_button;
*display_strings   = *main::display_strings;


#########################################################################
## Main Dispatcher
sub execute_mode {
        my $mode = shift;
        ({  
	    tables     =>  \& show_tables      ,
	    views      =>  \& show_views       ,
            databases  =>  \& show_databases   ,
            vacuum     =>  \& show_vacuum      ,
            stats      =>  \& show_stats       ,
            buffers    =>  \& show_buffers     ,
            indexes    =>  \& show_indexes     ,
            procedures =>  \& show_procedu     ,
            rules      =>  \& show_rules       ,
            settings   =>  \& show_settings    ,
            bucardo    =>  \& show_bucardo     ,
            dict       =>  \& show_dict        ,
            triggers   =>  \& show_triggers    ,
            users      =>  \& show_users       ,
         } -> {$mode})->();
}

#########################################################################

sub update_schema_display {
        #  Schema Table
        $::she     = get_schemas2( $opt, $::dbname) or return;
	my $val;
        eval { $val=0; $val = $::schemas->getField('VALUE')  };
        $::schemas = secondary_listbox('Schemas', $::she, 2,37, $val);
        $::schemas->execute($::mwh,0);
        ($::sname) = first_word( $::she->[ $::schemas->getField('VALUE')] );
}

sub update_big_display {
        # Result Table ( like relevant tables, indexes, objects, etc,. )
       ($::desc, $::actual) = @_ ;
        $::desc    = $::desc->();
        $::tab     = $::actual->( $opt, $::dbname, $::sname, $::secname);
	my $val;
        eval { $val=0; $val = $::big->getField('VALUE')  };
        $::big     = big_listbox( $::desc, $::tab, 11, 0, $val);
        $::big->execute($::mwh,0);
}
sub disable_schema_display {
	return unless $::schemas;
	my $misc = misc_system_wide($opt);
	$::schemas->setField( LISTITEMS =>$misc, CAPTION=>'' );
	$::schemas->draw($::mwh);
}
sub big_display_only { 
	disable_schema_display ; 
        goto &update_big_display;
}
sub whole_movie {
        ($::desc, $::actual) = @_ ;
	update_schema_display( $::desc, $::actual );
	update_big_display( $::desc, $::actual) ;
}
sub update_section_display {
	my ($title, $choices) = @_ ;
	$::she = $choices;
	my $val;
        eval { $val=0; $val = $::schemas->getField('VALUE')  };
        $::schemas  = secondary_listbox( $title, $::she, 2,37,$val);
        $::schemas->execute($::mwh,0);
        ($::secname) = first_word( $::she->[ $::schemas->getField('VALUE')]);
}

sub show_settings { 
        my $choi = [qw( All backend internal postmaster 
                        sighup superuser user changed)];
	update_section_display ('Context', $choi);
        update_big_display( sub{''}, \& all_settings);
}
sub show_buffers { 
        my $choi = [ 'user', 'system & user', 'not_cached'];
	update_section_display ('', $choi);
        update_big_display( \& pgbuffercache_desc, \& bufferca_);
}
sub bufferca_{ 
	pgbuffercache $opt, $::dbname, $::secname;
}
sub show_databases{ big_display_only( \& all_databases_desc,\& all_databases) } 
sub show_bucardo  { big_display_only( \& bucardo_conf_desc, \& bucardo_conf)} 
sub show_users    { big_display_only( \&get_users_desc, \&get_users     )  }

sub show_stats   { whole_movie( \&table_stats_desc, \&table_stats       )  }
sub show_dict    { whole_movie( \&dict_desc, \&dict                     )  }
sub show_tables  { whole_movie( \&tables_brief_desc, \&tables_brief     )  }
sub show_views   { whole_movie( \&get_views_all_desc, \&get_views_all   )  }
sub show_procedu { whole_movie( \&get_proc_desc, \&get_proc             )  }
sub show_indexes { whole_movie( \&index3_desc, \&index3                 )  }
sub show_rules   { whole_movie( \&rules_desc, \&rules                   )  }
sub show_triggers{ whole_movie( \&schema_trg_desc, \&schema_trg         )  }
sub show_vacuum  { whole_movie( \&tables_vacuum_desc, \&tables_vacuum   )  }


## Another dispatcher
sub retrieve_permit {
        my ($sna) = first_word( $::she->[ $::schemas->getField('VALUE')] );
	get_nspacl($opt, $::dbname, $sna) ;
}

## Another dispatcher
sub retrieve_context {
	#return if $::mode eq 'rules' ;
        ({  
	    tables     => \& tstat    ,
            views      => \& viewof   ,
            vacuum     => \& vacuumof ,
            databases  => \& over2    ,
            stats      => \& statsof  ,
            settings   => \& settingof,
            procedures => \& procof   ,
	    buffers    => \& bufferca ,
	    rules      => \& ruleof   ,
            indexes    => \& indexof  ,
            bucardo    => \& bucardoof,
            triggers   => \& trgof    ,
            users      => \& userof   ,

         }->{$::mode||return})->(@_) ;
}

### The following functions are colled from the above dispatcher
sub stats2 { [table_stats2_desc,  @{table_stats2($opt, $::dbname, $::sname )}]}
sub bufferca{ 
	[ @{buffercache_summary( $opt, $::dbname)} ,
	  '','',
	  #@{ pgbuffercache( $opt, $::dbname)} ,
	  @{ pgbufpages( $opt, $::dbname, $::secname)} ,
        ]
}

sub vacuumof {
        my $index = $::big->getField('VALUE');
        my ($f)   = first_word( $::tab->[$index] );
        vacuum_per_table( $opt, $::dbname, $::sname, $f ) or return [];
        #vacuum_per_table( $opt, $::dbname, split(/\./, $f,2) ) or return [];
}
sub userof {
        my $index = $::big->getField('VALUE');
        my ($f)   = first_word( $::tab->[$index] );
        user_of( $opt, $f ) or return [];
}
sub trgof {
        my $index = $::big->getField('VALUE');
        my ($l)   = last_word( $::tab->[$index] );
        trg_of( $opt, $::dbname, $::sname, $l ) or return [];
}

sub ruleof {
        my $index = $::big->getField('VALUE');
        my ($f)   = first_word( $::tab->[$index] );
        my $text  = rule_of( $opt, $::dbname, $::sname, $f ) or return [];
}

sub over2  { 
        my $index = $::big->getField('VALUE');
        my ($f)   = first_word( $::tab->[$index] );
        over_dbs( $opt, $f) ;
} 
sub bucardoof  { 
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
        bucardo_conf_of( $opt, $f) ;
} 


sub statsof  { 
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
        statsoftable( $opt, $::dbname, $::sname,  $f) ;
} 
sub sql_formater {
	local $_ = shift or return '';
	return $_  if /^(select|from)/i ;
	return "    $_";
}
sub viewof  { 
        my $index =  $::big->getField('VALUE');
        my ($f)   =  first_word( $::tab->[$index] );
        my $text  =  view_of( $opt, $::dbname, $::sname,  $f) ;
	[  '', 
	   map { sql_formater( $_ ) }
	   map { my @parts  = split  /\bFROM\b/i, $_ ,2  ;
		 (@parts>1) ?  ($parts[0], 'FROM '. $parts[1])  : $parts[0] 
                }
		Curses::Widgets::textwrap($text, 50) 
        ];
} 

sub settingof {
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
        get_setting( $opt, $f ) or return [];
}

sub last_word { local $_ = shift; my ($last) =  / \w+$/xg; $last; }
sub over3  { 
        my $index = $::big->getField('VALUE');
        my ($f)   = first_word( $::tab->[$index] );
        over_dbs3( $opt, $f) ;
} 

sub procof {
        my $index = $::big->getField('VALUE')   ; 
        my $last  = last_word( $::tab->[$index] )   or return ['selection?'] ;
        proc_of( $opt, $::dbname, $last );
}

sub indexof {
        my $index = $::big->getField('VALUE')   ; 
        my $last  = last_word( $::tab->[$index] )   or return ['selection?'] ;
        get_index( $opt, $::dbname, $last ) or return [];
}
sub tstat {
        my $index = $::big->getField('VALUE');
        my ($f)   = first_word( $::tab->[$index] );
        table_stat( $opt, $::dbname, $::sname, $f ) or return [];
}

##########################################################################
## Another dispatcher
sub capital_context {
        eval{
        ({  tables     => \& tdataof  ,
            stats      => \& tdataof  ,
            indexes    => \& idxdef   ,
            rules      => \& ruleof   ,
            views      => \& viewof   ,
            stats      => \& mostof   ,
            settings   => \& fsmvals  ,
            databases  => \& dbof     ,
            buffers    => \& pgbufall ,
            vacuum     => \& v_settings ,
         }->{$::mode||return})->(@_) ;
	 };
}

sub v_settings{ vac_settings $opt }
sub pgbufall{ pgbuff_all $opt, $::dbname, $::secname }
sub bufcalc { 
	     [ '',
               @{buffercache_summary( $opt, $::dbname)} ,
	       '',
               @{bufstat( $opt, $::dbname, $::sname)},
	     ]
}

sub idxdef {
        my $index = $::big->getField('VALUE');
        my ($l) = last_word( $::tab->[$index] );
        indexdef( $opt, $::dbname, $l ) or return [];
}
sub idx3b {
        my $index = $::big->getField('VALUE');
        my ($l) = last_word( $::tab->[$index] );
        index3b( $opt, $::dbname, $l ) or return [];
}
sub stat_of {
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
        statisticsof( $opt, $::dbname, $::sname, $f ) or return [];
}
sub rewrite_of {
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
        rewriteof( $opt, $::dbname, $::sname, $f ) or return [];
}
sub mostof {
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
        most_common( $opt, $::dbname, $::sname, $f ) or return [];
}
sub tdataof {
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
        tbl_data_of( $opt, $::dbname, $::sname, $f ) or return [];
}
sub dbof {
        my $index  = $::big->getField('VALUE');
        my ($f)    = first_word( $::tab->[$index] );
        my $title  = object_totals_desc;
        my @desc   = object_totals_desc_long;
        my $r1     = object_totals( $opt, $f, 'all'     );
        my $r2     = object_totals( $opt, $f, 'shared'  );
        my $r3     = object_totals( $opt, $f, 'noshared');
        my $r4     = object_totals( $opt, $f, 'noshared-nosystem');
	[ $f, '', $title, '' , 
          "@$r1" . "\t\t totals", 
          "@$r2" . "\t\t shared", 
          "@$r3" . "\t\t not shared", 
          "@$r4" . "\t\t not shared, non-system", 
          @desc,
        ];
}
sub fsmvals {
        fsm_settings( $opt, $::dbname ) or return [];
}

##########################################################################
## Other dispatchers

sub display_keyword {
        my $keyword = shift||return;
        my ($y,$x)  = (9,1) ;
        $::mwh->addstr( $y,$x, $keyword);
        $::mwh->refresh;
        sleep 1;
        $::mwh->addstr( $y,$x, ' ' x length$keyword);
        $::mwh->refresh;
}
sub update_big_d {
        $::tab     = tables_brief( $opt, $::dbname, $::sname);
	$::big->setField( LISTITEMS => $::tab );
}

sub update_bigbox_inc {
        $::tab = ({ tables    => \& tables_brief,
                    stats     => \& table_stats,
                    #vacuum    => \& tables_of_db,
                    vacuum    => \& tables_vacuum,
                    databases => \& all_databases,
                 } -> {$::mode})->($opt, $::dbname, $::sname);
	$::big->setField( LISTITEMS => $::tab );
}
sub analyze  {
        return unless $::mode =~ qr/^ (tables|stats|databases|vacuum) $/xo;
        ({ tables    => \& do_analyze_tbl,
           stats     => \& do_analyze_tbl,
           vacuum    => \& do_analyze_tbl,
           databases => \& do_analyze_db,
        } -> {$::mode})->() ? display_keyword 'ANALYZE'
                            : display_keyword 'failed';
	update_bigbox_inc ;
}

sub vacuum  {
        return unless $::mode =~ qr/^ (tables|stats|databases|vacuum) $/xo;
        ({ tables    => \& do_vacuum_tbl,
           stats     => \& do_vacuum_tbl,
           vacuum    => \& do_vacuum_tbl,
           databases => \& do_vacuum_db,
        } -> {$::mode})->() ? display_keyword 'VACUUM' 
                            : display_keyword 'failed' ;
	update_bigbox_inc ;
}

sub reindex  {
	return unless $::mode eq 'indexes';
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
	my $ret = ($::mode eq 'tables') 
		?  reindex_tbl( $opt, $::dbname, $::sname, $f )
		:  reindex_db( $opt, $::dbname ) ;
        ($ret)  ? display_keyword 'REINDEX' 
                : display_keyword 'failed' ;
}

sub do_analyze_tbl {
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
        analyze_tbl( $opt, $::dbname, $::sname, $f ) or return ;
}
sub do_analyze_db {
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
        #analyze_db( $opt, $::dbname ) or return ;
        analyze_db( $opt, $f ) or return ;
}
sub do_vacuum_tbl {
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
        vacuum_tbl( $opt, $::dbname, $::sname, $f ) or return ;
}
sub table2of {
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
        table2_of( $opt, $::dbname, $::sname, $f ) or return ;
}
sub table3of {
        my $index = $::big->getField('VALUE');
        my ($f) = first_word( $::tab->[$index] );
	my $nsp = $::sname;
	($f=~/\w\.\w/ )  and ($nsp,$f) = split /\./, $f ;
        table3_of( $opt, $::dbname, $nsp, $f ) or return ;
}
sub do_vacuum_db {
        my $index = $::big->getField('VALUE');
        my ($f)   = first_word( $::tab->[$index] );
        #vacuum_db( $opt, $::dbname ) or return ;
        vacuum_db( $opt, $f ) or return ;
}
sub save2file {
	eval { open my ($i) , '>>/tmp/pcurse.out';
	       print {$i} $::desc                ;
	       print {$i} $_  for @{$::tab}      ; 1;
	} or return;
        display_keyword 'SAVED ' ;
}

sub  display_button {
	my ($choices) = @_ ;
        $::butt       = create_button( $choices, 4, 58, 8);
        $::butt->execute($::mwh,0);
}
1;
__END__
=head1 NAME

Pg::Pcurse - Monitors a Postgres cluster


=head1 SYNOPSIS

  use Pg::Pcurse;

=head1 DESCRIPTION

Library functions for pcurse(1) which monitors Postgres databases



=head1 SEE ALSO

pcurse(1)


=head1 AUTHOR

Ioannis Tambouras, E<lt>ioannis@cpan.org<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by Ioannis Tambouras

This library is free software; you can redistribute it and/or modify
it under the same terms of GPLv3


=cut
